src/settings/toml/migrations.rs

```diff
@@ -1,44 +1,175 @@
+                ..
+            } => anyhow::bail!("You provided an existing migration tag with your adhoc migration but the script does not have a migration tag set."),
+            Migrations::Adhoc {
+                script_tag: MigrationTag::NoScript,
+                provided_old_tag: Some(_),
+                ..
+            } => anyhow::bail!("You provided an existing migration tag with your adhoc migration but the script does not exist."),
+            Migrations::Adhoc {
+                script_tag: MigrationTag::HasTag(existing_tag),
+                provided_old_tag: None,
+                ..
+            } => anyhow::bail!("You didn't provide an existing tag with your adhoc migration, but the script currently has a migration tag set to \"{}\"", existing_tag),
+            Migrations::Adhoc {
+                provided_old_tag: Some(_),
                 new_tag: None,
-                migration: migrations[0].migration.clone(),
-            }),
-            migrations => Self::coalesce_migrations(migrations),
-        }
+                ..
+            } => anyhow::bail!("You must provide a --new-tag when --old-tag is set."),
+            Migrations::Adhoc {
+                script_tag: MigrationTag::Unknown, ..
+            } => anyhow::bail!("The migration tag for the script was never fetched. This is a bug in wrangler, and we'd appreciate an issue report!"),
+            Migrations::List {
+                script_tag: MigrationTag::Unknown, ..
+            } => anyhow::bail!("The migration tag for the script was never fetched. This is a bug in wrangler, and we'd appreciate an issue report!"),
+            _ => ()
+        };
+
+        // validate migration list
+        if let Migrations::List { migrations, .. } = &self {
+            let mut seen = HashSet::new();
+            for migration in migrations {
+                if seen.contains(&migration.tag) {
+                    anyhow::bail!("The migration tag \"{}\" appears multiple times in the list of migrations in your wrangler.toml", migration.tag);
+                }
+                seen.insert(&migration.tag);
+            }
+        };
+
+        Ok(())
     }
 
-    fn coalesce_migrations(_migrations: &[MigrationConfig]) -> Result<ApiMigration, anyhow::Error> {
-        unimplemented!()
+    pub fn api_migration(&self) -> Result<Option<ApiMigration>, anyhow::Error> {
+        self.validate()?;
+        match &self {
+            Migrations::Adhoc {
+                provided_old_tag,
+                new_tag,
+                migration,
+                ..
+            } => {
+                let api_migration = ApiMigration {
+                    old_tag: provided_old_tag.to_owned(),
+                    new_tag: new_tag.to_owned(),
+                    steps: migration
+                        .as_ref()
+                        .map_or_else(Vec::new, |m| vec![m.clone()]),
+                };
+
+                log::info!("Api Migration: {:#?}", api_migration);
```

Comment:
Same question about whether this was intentional. If it was, I'd capitalize API.

Ditto below in the ::List case as well.